[MISC] Add Multistate Chip

acrowthe opened the pull request:
Add a chip that can be clicked to cycle between multiple different values and appearances

Here's an example test snippet, I pasted it at the top of live table but putting it into any component that's easy to access should work for testing:
```
      <TriStateChip
        label="Test Chip"
        defaultTooltip="Include/Exclude"
        positiveTooltip="Include"
        onChange={(value) => console.log(value)}
      />
      
```

acrowthe committed:
[MISC] Add Multistate Chip
Add a chip that can be clicked to cycle between multiple different values and appearances

veronikaslc commented on modules/data-entry/src/main/frontend/src/dataHomepage/MultiStateChip.jsx:
```diff
@@ -0,0 +1,99 @@
+  const [ currentStateIndex, setCurrentState ] = useState(0);
+
+  useMemo(() => {
+    if (currentStateIndex > states?.length && currentStateIndex > 0) {
```
I am confused here with the reason for adding the second part of expression. Is there a case the `currentStateIndex > states?.length` but can be `currentStateIndex <= 0`?

marta- replied:
Indeed the second part seems unnecessary. Any value less than 1 will make the first part of the expression false regardless of what is in `states`.

veronikaslc commented on modules/data-entry/src/main/frontend/src/dataHomepage/MultiStateChip.jsx:
```diff
@@ -0,0 +1,99 @@
+//  specific language governing permissions and limitations
+//  under the License.
+//
+import React, { useState, useMemo } from "react";
```
No `React` import after CARDS-2794

veronikaslc commented on modules/data-entry/src/main/frontend/src/dataHomepage/MultiStateChip.jsx:
```diff
@@ -0,0 +1,99 @@
+    }
+  }, [states])
+
+  return <Tooltip title={states[currentStateIndex]?.tooltip}>
```
I see we heavily rely on `states[currentStateIndex]?`. What if there is no `states[currentStateIndex]` or no `states`? Should we still attempt to render anything in this case case or abort early?

Also I'd add just to avoid code duplication 
`const state = states[currentStateIndex];`

marta- replied:
We should just make `states` required and ensure it's non-empty. There's no point in a multistate chip without having the states defined.

veronikaslc commented on modules/data-entry/src/main/frontend/src/dataHomepage/MultiStateChip.jsx:
```diff
@@ -0,0 +1,99 @@
+        onClick={() => {
+          let newIndex = (currentStateIndex + 1 >= states?.length) ? 0 : (currentStateIndex + 1);
+          setCurrentState(newIndex);
+          onChange(states[newIndex].value);
```
If we are not sure `states` exists here (from line 79 `states?.length`) `states[newIndex]` can thow error, also if we are generally not sure `states[currentStateIndex]` exists `states[newIndex].value` can also potentially throw error

marta- replied:
We must be sure that `states` is well defined before we start rendering.

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/MultiStateChip.jsx:
```diff
@@ -0,0 +1,99 @@
+        label={states[currentStateIndex]?.label}
+        variant={states[currentStateIndex]?.variant}
+        color={states[currentStateIndex]?.color}
+        icon={states[currentStateIndex]?.icon}
```
> Also I'd add just to avoid code duplication
> `const state = states[currentStateIndex];`

Or, a less verbose way of using the state props is to define the ChipState visual props as a plain object and expand it in the rendered `Chip`'s props:

```suggestion
        {...states[currentStateIndex]?.chipProps ?? {}} // this would contain label, variant, color, icon
```

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/MultiStateChip.jsx:
```diff
@@ -0,0 +1,99 @@
+    "states": states,
+    "onChange": onChange,
+  });
+}
```
This component should be in a separate file.

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/MultiStateChip.jsx:
```diff
@@ -0,0 +1,99 @@
+        color={states[currentStateIndex]?.color}
+        icon={states[currentStateIndex]?.icon}
+        onClick={() => {
+          let newIndex = (currentStateIndex + 1 >= states?.length) ? 0 : (currentStateIndex + 1);
```
```suggestion
          let newIndex = (currentStateIndex + 1) % states.length);
```
Must make sure that we don't get to this point with undefined or empty `states`.

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/MultiStateChip.jsx:
```diff
@@ -0,0 +1,99 @@
+    color: PropTypes.string,
+    icon: PropTypes.node,
+    value: PropTypes.string,
+  }))
```
```suggestion
  })).isRequired
```

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/MultiStateChip.jsx:
```diff
@@ -0,0 +1,99 @@
+    }
+  }, [states])
+
+  return <Tooltip title={states[currentStateIndex]?.tooltip}>
```
Only render the chip if we have states. Also there will be a warning if for any reason the tooltip is missing, are we making tooltips required, should we allow them to be absent and not render the tooltip in that case (tedious!) or should we us some sort of default?

```suggestion
  return states?.length > 0 &&
    <Tooltip title={states[currentStateIndex]?.tooltip ?? "Click to change"}>
```

acrowthe replied:
I'm not seeing any sort of error if tooltip gets a null or undefined title - it just silently displays nothing. I've flagged the labels as required props anyways

acrowthe committed:
[MISC] Add Multistate Chip
Code cleanup and improved error cases
- Move tri state chip into seperate component file
- Improve handling of cases where data is missing eg. undefined states
- Improve property definition to simplify passing properties down to chip

marta- reviewed (commented):
Looks clean and works well.

Needed change: move both files from `dataHomepage` to `commons`.

Nice to have change: for `TristateChip`, I would consider specific callbacks for switching between the 3 states. We currently have a generic `onChange`, but the implementation might look ugly if we need to check the state value each time. Instead, I would implement onChange internally:

```
if (newState == 0) onUnset();
else if (newState == 1) onSetPositive();
else onSetNegative();
```

and only expose these 3 new handlers. Better names for the `on___` functions are welcome as always, as is proper code style.
